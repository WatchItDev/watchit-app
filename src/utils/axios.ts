import axios, { AxiosRequestConfig } from 'axios';
// config
import { HOST_API } from 'src/config-global';

// ----------------------------------------------------------------------

const axiosInstance = axios.create({ baseURL: HOST_API });

axiosInstance.interceptors.response.use(
  (res) => res,
  (error) => Promise.reject((error.response && error.response.data) || 'Something went wrong')
);

export default axiosInstance;

// ----------------------------------------------------------------------

export const fetcher = async (args: string | [string, AxiosRequestConfig]) => {
  const [url, config] = Array.isArray(args) ? args : [args];

  const res = await axiosInstance.get(url, { ...config });

  return res.data;
};

// ----------------------------------------------------------------------

export const endpoints = {
  chat: '/api/chat',
  kanban: '/api/kanban',
  calendar: '/api/calendar',
  auth: {
    me: '/api/auth/me',
    login: '/api/auth/login',
    register: '/api/auth/register',
  },
  mail: {
    list: '/api/mail/list',
    details: '/api/mail/details',
    labels: '/api/mail/labels',
  },
  post: {
    list: '/api/post/list',
    details: '/api/post/details',
    latest: '/api/post/latest',
    search: '/api/post/search',
  },
  product: {
    list: '/api/product/list',
    details: '/api/product/details',
    search: '/api/product/search',
  },
};
