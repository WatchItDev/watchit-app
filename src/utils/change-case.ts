export function paramCase(str: string) {
  return str
    .toLowerCase()
    .replace(/\s+/g, '-')
    .replace(/[^a-z0-9-]/g, '');
}

export function snakeCase(str: string) {
  return str
    .toLowerCase()
    .replace(/\s+/g, '_')
    .replace(/[^a-z0-9_]/g, '');
}
