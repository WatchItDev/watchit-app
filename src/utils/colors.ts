/**
 * Adjusts the brightness of a given color by darkening it.
 * The function decreases the RGB components of the color to create a darker shade.
 *
 * @param {string} color - A string representing the hexadecimal color value (e.g., "#RRGGBB").
 * @param {number} [tone=30] - The amount to darken each RGB component. Default is 30.
 * @returns {string} A new hexadecimal color string that represents the darker shade of the input color.
 */
export const darken = (color: string, tone = 30) => {
  let r = parseInt(color.slice(1, 3), 16);
  let g = parseInt(color.slice(3, 5), 16);
  let b = parseInt(color.slice(5, 7), 16);

  r = Math.max(0, r - tone);
  g = Math.max(0, g - tone);
  b = Math.max(0, b - tone);

  return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
};
