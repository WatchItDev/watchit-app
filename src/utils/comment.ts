export const timeAgo = (date) => {
  const now = new Date();
  const seconds = Math.floor((now - date) / 1000);

  let interval = Math.floor(seconds / 31536000);
  if (interval >= 1) {
    return interval + 'y ago';
  }
  interval = Math.floor(seconds / 2592000);
  if (interval >= 1) {
    return interval + 'mo ago';
  }
  interval = Math.floor(seconds / 86400);
  if (interval >= 1) {
    return interval + 'd ago';
  }
  interval = Math.floor(seconds / 3600);
  if (interval >= 1) {
    return interval + 'h ago';
  }
  interval = Math.floor(seconds / 60);
  if (interval >= 1) {
    return interval + 'min ago';
  }
  return Math.floor(seconds) + 's ago';
};
