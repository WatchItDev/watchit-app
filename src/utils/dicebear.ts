export const dicebear = (seed: string) => {
  return `https://api.dicebear.com/9.x/bottts-neutral/svg?seed=${seed}`
}
