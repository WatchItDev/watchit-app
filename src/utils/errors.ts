/**
 * Enum of our known error names.
 */
export enum ERRORS {
  NOT_LOGGED_IN = 'NOT_LOGGED_IN',
  BUNDLER_UNAVAILABLE = 'BUNDLER_UNAVAILABLE',
  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
}

/**
 * Error messages dictionary mapping our ERROR_NAMES to user-facing strings.
 */
export const ERROR_MESSAGES: Record<ERRORS, string> = {
  [ERRORS.NOT_LOGGED_IN]: 'Please log in to continue.',
  [ERRORS.BUNDLER_UNAVAILABLE]: 'The bundler is currently unavailable.',
  [ERRORS.UNKNOWN_ERROR]: 'An unknown error has occurred.',
};

let globalEnqueueSnackbar: ((message: string, options?: object) => void) | null = null;

/**
 * Allows us to set the globalEnqueueSnackbar function from our App (or any top-level component).
 * Must be called at least once (e.g., in App.tsx) so that notifyError() works.
 */
export function setGlobalErrorNotifier(
  enqueueSnackbarFn: (message: string, options?: object) => void
) {
  globalEnqueueSnackbar = enqueueSnackbarFn;
}

/**
 * Global function to notify an error by ERROR_NAMES.
 * If no message is found in the dictionary, it uses 'UNKNOWN_ERROR'.
 */
export function notifyError(errorName: ERRORS, fallbackMessage?: string) {
  if (!globalEnqueueSnackbar) {
    console.error('No globalEnqueueSnackbar is set. Cannot notify error.');
    return;
  }
  const message =
    ERROR_MESSAGES[errorName] ||
    fallbackMessage ||
    ERROR_MESSAGES[ERRORS.UNKNOWN_ERROR];

  globalEnqueueSnackbar(message, { variant: 'error' });
}
