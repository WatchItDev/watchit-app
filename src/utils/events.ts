import { Dispatch } from 'redux'
import { appendNotification } from '@src/redux/notifications'

export namespace Events {
  export const Handlers = (payload: any, profileId: string, dispatch?: Dispatch) => {
    const evenType: keyof typeof EventsHandlersMap = payload?.new?.payload?.type

    if (EventsHandlersMap[evenType]) {
      EventsHandlersMap[evenType](payload, profileId, dispatch)
    }
  }

  /**
   * Handle notification event for internal popover.
   * @param payload Data from the event (set as needed)
   * @param profileId Current user profile id for LENS Protocol
   * @param dispatch Redux dispatch function
   * @constructor
   */
  export const Notification = (payload: any, profileId: string, dispatch?: Dispatch) => {
    const notification = payload.new

    if (notification?.receiver_id === profileId) {
      if (dispatch) {
        dispatch(appendNotification(notification))
      }
    }
  }
  // @ts-ignore
  export const Transaction = (payload: any) => {}
  // @ts-ignore
  export const Publication = (payload: any) => {}
  // @ts-ignore
  export const Subscription = (payload: any) => {}
  // @ts-ignore
  export const Follower = (payload: any) => {}
}

export const EventsHandlersMap = {
  NOTIFICATION: Events.Notification,
  TRANSACTION: Events.Transaction,
  PUBLICATION: Events.Publication,
  SUBSCRIPTION: Events.Subscription,
  FOLLOWER: Events.Follower,
}
