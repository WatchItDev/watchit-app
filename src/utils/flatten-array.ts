// ----------------------------------------------------------------------

export function flattenArray<T>(list: T[], key = 'children'): T[] {
  let children: T[] = [];

  const flatten = list?.map((item: any) => {
    if (item[key] && item[key].length) {
      children = [...children, ...item[key]];
    }
    return item;
  });

  return flatten?.concat(children.length ? flattenArray(children, key) : children);
}
