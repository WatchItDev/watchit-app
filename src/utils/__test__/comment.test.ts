import { timeAgo } from '@src/utils/comment';

describe('[UTILS]: timeAgo', () => {
  it('displays time in seconds if less than a minute', () => {
    expect(timeAgo(new Date(Date.now() - 30 * 1000))).toBe('30s ago');
  });

  it('displays time in minutes if less than an hour', () => {
    expect(timeAgo(new Date(Date.now() - 5 * 60 * 1000))).toBe('5min ago');
  });

  it('displays time in hours if less than a day', () => {
    expect(timeAgo(new Date(Date.now() - 3 * 60 * 60 * 1000))).toBe('3h ago');
  });

  it('displays time in days if less than a month', () => {
    expect(timeAgo(new Date(Date.now() - 5 * 24 * 60 * 60 * 1000))).toBe('5d ago');
  });

  it('displays time in months if less than a year', () => {
    expect(timeAgo(new Date(Date.now() - 2 * 30 * 24 * 60 * 60 * 1000))).toBe('2mo ago');
  });

  it('displays time in years if more than a year', () => {
    expect(timeAgo(new Date(Date.now() - 2 * 365 * 24 * 60 * 60 * 1000))).toBe('2y ago');
  });

  it('matches the snapshot for various time ranges', () => {
    const now = Date.now();
    expect([
      timeAgo(new Date(now - 10 * 1000)),        // seconds
      timeAgo(new Date(now - 2 * 60 * 1000)),    // minutes
      timeAgo(new Date(now - 4 * 3600 * 1000)),  // hours
      timeAgo(new Date(now - 3 * 86400 * 1000)), // days
      timeAgo(new Date(now - 2 * 2592000 * 1000)), // months
      timeAgo(new Date(now - 1.5 * 31536000 * 1000)) // years
    ]).toMatchSnapshot();
  });
});
