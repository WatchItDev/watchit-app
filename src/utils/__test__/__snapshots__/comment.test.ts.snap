// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[UTILS]: timeAgo > matches the snapshot for various time ranges 1`] = `
[
  "10s ago",
  "2min ago",
  "4h ago",
  "3d ago",
  "2mo ago",
  "1y ago",
]
`;
